File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/RewardVideo/AlxTopOnRewardVideoAdapter.m
//
//  AlxTopOnRewardVideoAdapter.m
//  AlxAdsOCDemo
//

#import "AlxTopOnRewardVideoAdapter.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnRewardVideoDelegate.h"
#import "AlxTopOnRewardVideoEvent.h"
#import "AlxTopOnBiddingRequest.h"
#import "AlxTopOnBiddingRequestManager.h"
#import "AlxTopOnTool.h"

static NSString *const TAG = @"AlxTopOnRewardVideoAdapter";

@interface AlxTopOnRewardVideoAdapter ()
@property (nonatomic, strong) AlxRewardVideoAd *rewardedAd;
@property (nonatomic, strong) AlxTopOnRewardVideoDelegate *rewardVideoDelegate;
@end

@implementation AlxTopOnRewardVideoAdapter

#pragma mark - lazy load
- (AlxTopOnRewardVideoDelegate *)rewardVideoDelegate {
    if (_rewardVideoDelegate == nil) {
        _rewardVideoDelegate = [[AlxTopOnRewardVideoDelegate alloc] init];
        _rewardVideoDelegate.adStatusBridge = self.adStatusBridge;
    }
    return _rewardVideoDelegate;
}

#pragma mark - Ad load
- (void)loadADWithArgument:(ATAdMediationArgument *)argument {
    NSLog(@"%@: loadADWithArgument", TAG);
    NSLog(@"%@: loadAD: isMainThread=%@", TAG, [NSThread isMainThread] ? @"YES" : @"NO");
    
    NSString *bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey];
    NSLog(@"%@: loadAD: bidId=%@", TAG, bidId ?: @"空");
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *unitKeyStr = [AlxTopOnBaseManager unitID];
        NSString *unitId = argument.serverContentDic[unitKeyStr];
        
        if (!unitId || unitId.length == 0) {
            NSString *errorStr = @"unitid is empty";
            NSLog(@"%@: loadAD: error = %@", TAG, errorStr);
            NSError *error = [AlxTopOnBaseManager errorWithCode:-100 message:errorStr];
            [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            return;
        }
        NSLog(@"%@: loadAD: unitid = %@", TAG, unitId);
        
        if (bidId) {
            // Bidding 场景：从缓存中取出已加载的广告 / Bidding scenario: retrieve the pre-loaded ad from cache
            AlxTopOnBiddingRequest *biddingRequest = [[AlxTopOnTool shared] getRequestItemWithUnitID:unitId];
            if (biddingRequest) {
                self.rewardedAd = (AlxRewardVideoAd *)biddingRequest.customObject;
                
                if (self.rewardedAd) {
                    NSLog(@"%@: loadAD: bid ad loaded, notify success", TAG);
                    // ⚠️ 注意：将广告对象传给 TopOn SDK / Note: pass the ad object to TopOn SDK
                    NSMutableDictionary *adExtra = [NSMutableDictionary dictionary];
                    adExtra[kATAdAssetsCustomObjectKey] = self.rewardedAd;
                    [self.adStatusBridge atOnRewardedAdLoadedExtra:adExtra];
                } else {
                    NSLog(@"%@: loadAD: bid ad object is empty", TAG);
                    NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid ad object is empty"}];
                    [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
                }
            } else {
                NSLog(@"%@: loadAD: bid request not found in cache", TAG);
                NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid request not found"}];
                [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            }
            [[AlxTopOnTool shared] removeRequestItemWithUnitID:unitId];
        } else {
            // 普通加载场景 / Normal loading scenario
            self.rewardedAd = [[AlxRewardVideoAd alloc] init];
            self.rewardedAd.delegate = self.rewardVideoDelegate;
            // 设置 Delegate 的广告对象引用 / Set the delegate's ad object reference
            self.rewardVideoDelegate.rewardedAd = self.rewardedAd;
            
            NSLog(@"%@: start loading ad with unitId: %@", TAG, unitId);
            [self.rewardedAd loadAdWithAdUnitId:unitId];
        }
    });
}

#pragma mark - C2S Bidding
+ (void)bidRequestWithPlacementModel:(ATPlacementModel *)placementModel
                      unitGroupModel:(ATUnitGroupModel *)unitGroupModel
                                info:(NSDictionary *)info
                          completion:(void (^)(ATBidInfo * _Nullable, NSError * _Nullable))completion {
    NSLog(@"%@: bidRequestWithPlacementModel", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (![AlxTopOnBaseManager isInitialized]) {
            [AlxTopOnBaseManager initSDKWithServerInfo:info];
        }
    });
    
    AlxTopOnRewardVideoEvent *customEvent = [[AlxTopOnRewardVideoEvent alloc] initWithInfo:info localInfo:info];
    customEvent.isC2SBiding = YES;
    
    AlxTopOnBiddingRequest *request = [[AlxTopOnBiddingRequest alloc]
                                       initWithUnitGroup:unitGroupModel
                                       customEvent:customEvent
                                       unitID:info[[AlxTopOnBaseManager unitID]]
                                       placementID:placementModel.placementID
                                       extraInfo:info
                                       adType:ATAdFormatRewardedVideo
                                       bidCompletion:completion];
    
    [[AlxTopOnBiddingRequestManager shared] startWithRequest:request];
}

#pragma mark - Ad Ready Check (实例方法 / Instance Method)
- (BOOL)adReadyRewardedWithInfo:(NSDictionary *)info {
    NSLog(@"%@: adReadyRewardedWithInfo", TAG);
    
    // 检查广告对象是否准备好 / Check if the ad object is ready
    if (self.rewardedAd) {
        // 可以调用 isReady 或其他检查方法 / Can call isReady or other check methods
        NSLog(@"%@: adReady = YES", TAG);
        return YES;
    }
    
    NSLog(@"%@: adReady = NO", TAG);
    return NO;
}

#pragma mark - Show Ad (实例方法 / Instance Method)
- (void)showRewardedVideoInViewController:(UIViewController *)viewController {
    NSLog(@"%@: showRewardedVideoInViewController", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.rewardedAd) {
            // 展示广告 / Show the ad
            [self.rewardedAd showAdWithPresent:viewController];
            NSLog(@"%@: ad shown", TAG);
        } else {
            NSLog(@"%@: rewardedAd is nil", TAG);
        }
    });
}

@end

File: custom_adapter/TopOnAdapter/Swift/6.5.x/RewardVideo/AlxTopOnRewardVideoAdapter.swift
//
//  AlxTopOnRewardVideoAdapter.swift
//

import Foundation
import AnyThinkSDK
import AlxAds

@objc(AlxTopOnRewardVideoAdapter)
public class AlxTopOnRewardVideoAdapter: AlxTopOnBaseAdapter, ATBaseRewardedAdapterProtocol {
    
    private static let TAG = "AlxTopOnRewardVideoAdapter"
    
    private var rewardedAd: AlxRewardVideoAd?
    private lazy var rewardVideoDelegate: AlxTopOnRewardVideoDelegate = {
        let delegate = AlxTopOnRewardVideoDelegate()
        delegate.adStatusBridge = self.adStatusBridge
        return delegate
    }()
    
    // MARK: - ATBaseRewardedAdapterProtocol
    
    @objc public var adStatusBridge: ATRewardedAdStatusBridge!
    
    // MARK: - Ad Load
    
    @objc public override func loadAD(with argument: ATAdMediationArgument) {
        NSLog("%@: loadAD(with:)", AlxTopOnRewardVideoAdapter.TAG)
        NSLog("%@: loadAD: isMainThread=%@", AlxTopOnRewardVideoAdapter.TAG, Thread.current.isMainThread ? "YES" : "NO")
        
        let bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey] as? String
        NSLog("%@: loadAD: bidId=%@", AlxTopOnRewardVideoAdapter.TAG, bidId ?? "空")
        
        DispatchQueue.main.async {
            let unitKeyStr = AlxTopOnBaseManager.unitID
            guard let unitId = argument.serverContentDic[unitKeyStr] as? String, !unitId.isEmpty else {
                let errorStr = "unitid is empty"
                NSLog("%@: loadAD: error = %@", AlxTopOnRewardVideoAdapter.TAG, errorStr)
                let error = AlxTopOnBaseManager.error(code: -100, msg: errorStr)
                self.notifyLoadFailed(error: error)
                return
            }
            NSLog("%@: loadAD: unitid = %@", AlxTopOnRewardVideoAdapter.TAG, unitId)
            
            if let bidId = bidId {
                // Bidding 场景：从缓存中取出已加载的广告 / Bidding scenario: retrieve the pre-loaded ad from cache
                if let biddingRequest = AlxTopOnTool.shared.getRequestItem(withUnitID: unitId) as? AlxTopOnBiddingRequest {
                    self.rewardedAd = biddingRequest.customObject as? AlxRewardVideoAd
                    
                    if let rewardedAd = self.rewardedAd {
                        NSLog("%@: loadAD: bid ad loaded, notify success", AlxTopOnRewardVideoAdapter.TAG)
                        var adExtra: [AnyHashable: Any] = [:]
                        adExtra[kATAdAssetsCustomObjectKey] = rewardedAd
                        self.notifyRewardedAdLoaded(adExtra: adExtra)
                    } else {
                        NSLog("%@: loadAD: bid ad object is empty", AlxTopOnRewardVideoAdapter.TAG)
                        let error = NSError(domain: "AlxTopOnAdapter", code: -100, userInfo: [NSLocalizedDescriptionKey: "Bid ad object is empty"])
                        self.notifyLoadFailed(error: error)
                    }
                } else {
                    NSLog("%@: loadAD: bid request not found in cache", AlxTopOnRewardVideoAdapter.TAG)
                    let error = NSError(domain: "AlxTopOnAdapter", code: -100, userInfo: [NSLocalizedDescriptionKey: "Bid request not found"])
                    self.notifyLoadFailed(error: error)
                }
                AlxTopOnTool.shared.removeRequestItem(withUnitID: unitId)
            } else {
                // 普通加载场景 / Normal loading scenario
                self.rewardedAd = AlxRewardVideoAd()
                self.rewardedAd?.delegate = self.getRewardVideoDelegate()
                self.rewardVideoDelegate.rewardedAd = self.rewardedAd
                
                NSLog("%@: start loading ad with unitId: %@", AlxTopOnRewardVideoAdapter.TAG, unitId)
                self.rewardedAd?.loadAd(adUnitId: unitId)
            }
        }
    }
    
    private func getRewardVideoDelegate() -> AlxTopOnRewardVideoDelegate {
        self.rewardVideoDelegate.adStatusBridge = self.adStatusBridge
        return self.rewardVideoDelegate
    }
    
    // MARK: - Dynamic Invocation Helper Methods
    
    private func notifyRewardedAdLoaded(adExtra: [AnyHashable: Any]) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnRewardedAdLoadedExtra:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: adExtra)
            }
        }
    }
    
    private func notifyLoadFailed(error: Error) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdLoadFailed:adExtra:")
            if bridge.responds(to: selector) {
                let emptyDict: [AnyHashable: Any] = [:]
                bridge.perform(selector, with: error, with: emptyDict)
            }
        }
    }
    
    // MARK: - C2S Bidding
    
    @objc public static func bidRequestWithPlacementModel(_ placementModel: ATPlacementModel,
                                                          unitGroupModel: ATUnitGroupModel,
                                                          info: [AnyHashable: Any],
                                                          completion: @escaping (ATBidInfo?, Error?) -> Void) {
        NSLog("%@: bidRequestWithPlacementModel", AlxTopOnRewardVideoAdapter.TAG)
        
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: info)
            }
        }
        
        let customEvent = AlxTopOnRewardVideoEvent(info: info, localInfo: info)
        customEvent.isC2SBiding = true
        
        let request = AlxTopOnBiddingRequest(
            unitGroup: unitGroupModel,
            customEvent: customEvent,
            unitID: info[AlxTopOnBaseManager.unitID] as? String,
            placementID: placementModel.placementID,
            extraInfo: info,
            adType: ATAdFormat.rewardedVideo,
            bidCompletion: completion
        )
        
        AlxTopOnBiddingRequestManager.shared.start(with: request)
    }
    
    // MARK: - Ad Ready Check (实例方法 / Instance Method)
    
    @objc public func adReadyRewarded(withInfo info: [AnyHashable: Any]) -> Bool {
        NSLog("%@: adReadyRewardedWithInfo", AlxTopOnRewardVideoAdapter.TAG)
        
        if self.rewardedAd != nil {
            NSLog("%@: adReady = YES", AlxTopOnRewardVideoAdapter.TAG)
            return true
        }
        
        NSLog("%@: adReady = NO", AlxTopOnRewardVideoAdapter.TAG)
        return false
    }
    
    // MARK: - Show Ad (实例方法 / Instance Method)
    
    @objc(showRewardedVideoInViewController:)
    public func showRewardedVideo(in viewController: UIViewController) {
        NSLog("%@: showRewardedVideoInViewController", AlxTopOnRewardVideoAdapter.TAG)
        
        DispatchQueue.main.async {
            if let rewardedAd = self.rewardedAd {
                rewardedAd.showAd(present: viewController)
                NSLog("%@: ad shown", AlxTopOnRewardVideoAdapter.TAG)
            } else {
                NSLog("%@: rewardedAd is nil", AlxTopOnRewardVideoAdapter.TAG)
            }
        }
    }
}
